archive/imo/imo1960_q1.lean

```diff
@@ -0,0 +1,190 @@
+  rw [nat.digits, nat.digits_aux_def _ _ _ h1],
+  unfold sum_of_squares,
+  simp [pow_two],
+end
+
+/-
+This part is doing an exhaustive search.
+-/
+
+def fails_sum (c : ℕ) := c ≠ sum_of_digit_squares (c*11)
+
+def multiples_of_11 (n : ℕ) (h1: problem_predicate n) :
+∃ c : ℕ, ¬ fails_sum c ∧ 9 < c ∧ c < 91 ∧ n = c * 11 :=
+begin
+  obtain ⟨c, h2⟩ : ∃ c : ℕ, n = c * 11, from exists_eq_mul_left_of_dvd h1.right.left,
+  refine ⟨c, _, _, _, _⟩,
+  { have h3: c = (c * 11) / 11, by simp,
+    have h4: c = n / 11, from h2.symm ▸ h3,
+    unfold fails_sum,
+    rw [h2.symm, h4],
+    simp,
+    exact h1.right.right },
+  { have h5: n ≥ 100, from ge_100 n h1,
+    linarith },
+  { have h6: n < 1000, from lt_1000 n h1,
+    linarith },
+  { exact h2 },
+end
+
+lemma step_helper (b c : ℕ) (h : b < c) : b + 1 = c ∨ b + 1 < c := eq_or_lt_of_le h
+
+lemma iterative_step (c bound : ℕ) (h1 : fails_sum (bound + 1)) (h2 : fails_sum c ∨ bound < c) :
+fails_sum c ∨ bound + 1 < c :=
+begin
+  cases h2,
+  { left,
+    exact h2 },
+  have h3 : bound + 1 = c ∨ bound + 1 < c, from step_helper bound c h2,
+  cases h3,
+  { left,
+    exact h3 ▸ h1 },
+  right,
+  exact h3,
+end
+
+lemma low_search (c : ℕ) (lower_bound: 9 < c) : fails_sum c ∨ 49 < c :=
+begin
+  iterate 40 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+lemma mid_search (c : ℕ) (lower_bound : 50 < c) : fails_sum c ∨ 72 < c :=
+begin
+  iterate 22 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+lemma high_search (c : ℕ) (lower_bound : 73 < c) : fails_sum c ∨ 90 < c :=
+begin
+  iterate 17 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+/-
+Now we just need to combine the results from the `search` lemmas.
+-/
+
+lemma right_direction (n : ℕ) : problem_predicate n → solution_predicate n :=
+begin
+  intro ppn,
+  have h1 : ∃ c : ℕ, ¬ fails_sum c ∧ 9 < c ∧ c < 91 ∧ n = c * 11, from multiples_of_11 n ppn,
+  cases h1 with c h2,
+  have h3 : fails_sum c ∨ 49 < c, from low_search c h2.right.left,
+  cases h3,
+  { exact absurd h3 h2.left },
+  have h4 : 50 = c ∨ 50 < c, from step_helper 49 c h3,
+  have h5 : n = c * 11, from h2.right.right.right,
+  cases h4,
+  { left,
+    linarith },
+  have h6 : fails_sum c ∨ 72 < c, from mid_search c h4,
+  cases h6,
+  { exact absurd h6 h2.left },
+  have h7 : 73 = c ∨ 73 < c, from step_helper 72 c h6,
+  cases h7,
+  { right,
+    linarith },
+  have h8 : fails_sum c ∨ 90 < c, from high_search c h7,
+  cases h8,
+  { exact absurd h8 h2.left },
+  linarith,
+end
+
+lemma left_direction (n : ℕ) : solution_predicate n → problem_predicate n :=
+begin
+  intro spn,
+  cases spn,
+  { rw spn,
+    norm_num [problem_predicate, sum_of_squares] },
+  rw spn,
+  norm_num [problem_predicate, sum_of_squares],
+end
```

Comment:
And if we follow @bryangingechen's style suggestion on the other lemma, moving the implication to become a hypothesis, it becomes 
```suggestion
lemma left_direction (n : ℕ) (spn : solution_predicate n) : problem_predicate n :=
by rcases spn with (rfl | rfl); norm_num [problem_predicate, sum_of_squares]
```

Reply:
I made a different change that was suggested, which turns this lemma into a one-liner in a different way. So I think this should be resolved now.

Reply:
This one is slightly preferable, according to mathlib style, which calls avoiding implications in the statement of the main goal of the lemma (instead you move them before the colon, to become hypotheses).  Maybe this is to help with `library_search`, to allow for permutations of the hypotheses?  Not sure.

Reply:
OK